### ISSUE TITLE
Question: How to author dependency properties that depend on each other?

### ISSUE THREAD

**Issue by ackh** [microsoft/microsoft-ui-xaml#2452]
I'm working on a C++/CX based user interface element and I'm stuck with a particular problem. I'm asking here because I cannot find documentation or sample code that would help. I've asked the same question on [Stackoverflow](https://stackoverflow.com/q/61804996) but it seems to be largely ignored by people. Apologies if this is not the right location but I'm frankly running out of options.

My C++/CX user interface element contains four dependency properties. Three of those properties set the color channels red, green and blue in the underlying renderer. The C++/CX code for one such property looks as follows:

    uint8_t DemoControl::Red::get()
    {
      return static_cast<uint8_t>(GetValue(RedProperty));
    }

    void DemoControl::Red::set(uint8_t r)
    {
      SetValue(RedProperty, r);
    }

    DependencyProperty^ DemoControl::_redProperty =
      DependencyProperty::Register("Red",
                                   uint8_t::typeid,
                                   DemoControl::typeid,
                                   ref new PropertyMetadata(127, ref new PropertyChangedCallback(&DemoControl::OnRedChanged)));
    
    void DemoControl::OnRedChanged(DependencyObject^ d, DependencyPropertyChangedEventArgs^ e)
    {
      DemoControl^ DemoControl = static_cast<DemoControl^>(d);
      DemoControl->renderer->SetRed(static_cast<uint8_t>(e->NewValue));
    }

The fourth property returns the entire color, i.e. it is a combination of the values of the three other properties.

The question is, how would I update that color property if either the red, green or blue property changes without triggering the code attached to the color property via data binding?

A similar question has been asked [here](https://stackoverflow.com/q/1364484) but for WPF. The answer suggests to use [value coercion](https://docs.microsoft.com/en-us/previous-versions/dotnet/netframework-4.0/ms745795(v=vs.100)?redirectedfrom=MSDN) but this seems to be a feature unavailable to Windows Runtime components. The `PropertyMetadata` object used when registering the dependency property does not support `CoerceValueCallback` from what I can see.

There's a second part to the above piece of code:
From what I read in the documentation it is necessary to use `uint8` instead of `uint8_t` in order to be able to expose the Red, Green and Blue properties as a Windows Runtime public API. The problem is that I have no clue what to put in `DependencyProperty::Register` as second argument. `uint8::typeid` does not work nor does anything else I've tried.

**Comment by ranjeshj** [comment]
@MikeHillberg as FYI.

**Comment by michael-hawker** [comment]
@ackh posted a suggestion on StackOverflow, hope that helps. For this specific scenario, I think you can just have a common callback for all three of your color value properties and use that to construct the color as needed.

In other cases, I've found using the [Behavior](https://github.com/Microsoft/XamlBehaviors) pattern to be useful when I need to have multiple values loaded before processing them. Though that doesn't always solve the update problem.

This does sound like a good area for the docs to be improved for these types of complex inter-dependent property scenarios with suggested guidance and examples.

**Comment by ackh** [comment]
Thanks for the reply on StackOverflow, @michael-hawker, really appreciated that you took the time.

Currently, I'm using separate callbacks to react on property changes. The combined color property callback updates the other three color channel properties if necessary. To prevent infinite loops it sets a boolean variable for each of the three color channels. In the callbacks of the three color property channels those booleans are checked and further execution of the callbacks is aborted if the booleans are set. If any of the three color channel properties gets modified, the corresponding callback will update the combined color property. The same approach with a boolean as before is used in such a case.

This approach works but I guess I was hoping for some mechanism provided by the binding system that has the same effect without requiring the booleans. I leave the StackOverflow question open for a bit to see if more ideas are coming in.

Regarding the second part of my question:

The type `uint8` is located the `default` namespace. Somehow the compiler does not resolve that namespace automatically when defining the `static DependencyProperty`. So, in order to make it compile I had to specify `TypeName(::default::uint8::typeid)` as second argument of `DependencyProperty::Register`.

However, as it eventually turned out byte properties are not supported in XAML. There seems to be no built-in type converter and custom type converters are also not available. So, I ended up changing `uint8` to `uint32`. Technically that works but its just not as nice as having a byte property that would automatically limit each color channel to 8-bit.

I think it makes sense to close my question. Sorry for the noise that it caused.
